lib/addons/karpenter/index.ts

```diff
@@ -0,0 +1,118 @@
+        super({...defaultProps, ...props});
+        this.options = this.props;
+    }
+
+    deploy(clusterInfo: ClusterInfo): void {
+        const endpoint = clusterInfo.cluster.clusterEndpoint
+        const name = clusterInfo.cluster.clusterName
+        const cluster = clusterInfo.cluster
+        const values = { ...this.props.values ?? {} }
+
+        // Set up Node Role
+        const karpenterNodeRole = new Role(cluster, 'karpenter-node-role', {
+            assumedBy: new ServicePrincipal(`ec2.${cluster.stack.urlSuffix}`),
+            managedPolicies: [
+                ManagedPolicy.fromAwsManagedPolicyName("AmazonEKSWorkerNodePolicy"),
+                ManagedPolicy.fromAwsManagedPolicyName("AmazonEKS_CNI_Policy"),
+                ManagedPolicy.fromAwsManagedPolicyName("AmazonEC2ContainerRegistryReadOnly"),
+                ManagedPolicy.fromAwsManagedPolicyName("AmazonSSMManagedInstanceCore"),
+            ],
+            roleName: `KarpenterNodeRole-${name}`
+        });
+
+        // Set up Instance Profile
+        const karpenterInstanceProfile = new CfnInstanceProfile(cluster, 'karpenter-instance-profile', {
+            roles: [karpenterNodeRole.roleName],
+            instanceProfileName: `KarpenterNodeInstanceProfile-${name}`,
+            path: '/'
+        });
+
+        // Map Node Role to aws-auth
+        cluster.awsAuth.addRoleMapping(karpenterNodeRole, {
+            groups: ['system:bootstrapper', 'system:nodes'],
+            username: 'system:node:{{EC2PrivateDNSName}}'
+        })
+
+        // Create Namespace & SA
+        const ns = this.createNamespace(clusterInfo)
+        const sa = this.createServiceAccount(clusterInfo)
+        sa.node.addDependency(ns)
+
+        // Add helm chart
+        setPath(values, "serviceAccount.create", false)
+        setPath(values, "controller.clusterEndpoint", endpoint)
+        setPath(values, "controller.clusterName", name)
+        const karpenterChart = this.addHelmChart(clusterInfo, values, true)
+
+        karpenterChart.node.addDependency(sa);
+    }
+
+    /**
+     * Creates namespace, which is a prerequisite for service account creation and subsequent chart execution.
+     * @param clusterInfo 
+     * @returns 
+    */
+     protected createNamespace(clusterInfo: ClusterInfo): KubernetesManifest {
+        return new KubernetesManifest(clusterInfo.cluster.stack, "karpenter-namespace-struct", {
+            cluster: clusterInfo.cluster,
+            manifest: [{
+                apiVersion: 'v1',
+                kind: 'Namespace',
+                metadata: {
+                    name: this.options.namespace,
+                }
+            }],
+            overwrite: true,
+            prune: true
+        });
+    }
+
+    /**
+     * Creates a service account that can access secrets
+     * @param clusterInfo 
+     * @returns sa
+     */
+     protected createServiceAccount(clusterInfo: ClusterInfo): ServiceAccount {
```

Comment:
Sounds like another handy utility for the project that other addons can benefit from  as well. Let's move to utils and make it accept IAM policy statements as a param.

Reply:
I used PolicyDocument (which is a grouping of policy statements) as a param.